README.md
## FixDart
Dart Fix Protocol 4.4 Client Library

### Usage

A simple usage example:

```dart
import 'package:fix_dart/fix_dart.dart';

void main() async {
  var client = FixDart(
      host: '0.0.0.0',
      port: 0000,
      username: '-',
      password: '-',
      senderCompID: '-',
      targetCompID: '-');

  client.connect(
    autoReconnect: false,
    onConnected: () {
      client.subscribeMarketData('EURUSD.spa');
    },
    onQuoteReceived: (quote) {
      print(quote.describe());
    },
  );
}


```

## Limitation
Current limitation of this library:
- Only support market data
- Only support FIX 4.4
- No SSL support


CHANGELOG.md
## 0.0.1+7
- added auto reconnect setting to class
- added onWriteLog callback
- added ability to subscribe market data with different depth

## 0.0.1+6
- fixing bugs parameter not included on auto reconnect

## 0.0.1+5
- add heartbeat message every 30s
## 0.0.1+4
- add access to connection

## 0.0.1+3
- fixed repeated group in single message

## 0.0.1+2
- Enabled X tag

## 0.0.1+1

- Enabled TCP_NODELAY for socket option

## 0.0.1

- Initial version
- FIX Logon
- FIX MarketDataRequest


test/fix_dart_test.dart
import 'package:fix_dart/fix_dart.dart';
import 'package:test/test.dart';

void main() {
  group('A group of tests', () {
    FixDart client;

    setUp(() {
      client = FixDart();
    });

    test('First Test', () {
      expect(client.isConnected, isTrue);
    });
  });
}


main.dart
import 'dart:convert';
import 'dart:io';

enum MsgTypeValue { MsgType, MsgType2 }

class FixTag {
  String tag;
  dynamic value;
  FixTag(this.tag, this.value);

  String toKV() => '$tag=$value';
}

class MsgType extends FixTag {
  MsgType(dynamic val) : super('35', val);
}

void main() {
  print((10).toString().padLeft(3, '0'));
  print(DateTime.now()
      .toUtc()
      .toIso8601String()
      .replaceAll('-', '')
      .replaceAll('T', '-')
      .replaceAll('Z', '')
      .substring(0, 21));

  // var date1 = '20201005-02:10:27.010'.replaceFirst('-', ' ');
  // var date2 = '20201002-14:08:14'.replaceFirst('-', ' ');
  // print(DateTime.parse('${date1}Z'));
  // print(DateTime.parse('${date2}Z'));

  var fix44 = json.decode(File('fix44.json').readAsStringSync());

  /* Create tag defenition*/
  // for (var key in fix44['tags'].keys) {
  //   var value = fix44['tags'][key];
  //   var type = value.runtimeType.toString();
  //   var isString = type == 'String';
  //   if (isString) {
  //     print(
  //         "class $key extends FixTag { $key(dynamic val) : super('$value', val);}");
  //   }
  // }

  /* Create MsgType enum*/
  var content = fix44['tags']['MsgType']['values'];
  var enumCode = 'enum MsgTypes { ${content.keys.join(',')} }';
  var enumConverter = [];
  for (var key in content.keys) {
    enumConverter.add("MsgTypes.$key:'${content[key]}'");
  }
  print(enumCode);
  print('var converter<MsgTypes,String> = {${enumConverter.join(',')}};');
}


resources/parsing.dart
void main(List<String> args) {
  var messages =
      '8=FIX.4.4|9=124|35=W|34=90|49=QUOTES|52=20201007-06:50:01.590|56=200171|55=EURAUD.spa|262=200171_4|268=2|269=0|270=1.64739|269=1|270=1.6475|10=088|8=FIX.4.4|9=125|35=W|34=91|49=QUOTES|52=20201007-06:50:01.590|56=200171|55=AUDUSD.spa|262=200171_3|268=2|269=0|270=0.71328|269=1|270=0.71336|10=125|8=FIX.4.4|9=123|35=W|34=92|49=QUOTES|52=20201007-06:50:01.590|56=200171|55=EURGBP.spa|262=200171_7|268=2|269=0|270=0.91091|269=1|270=0.911|10=021|8=FIX.4.4|9=125|35=W|34=93|49=QUOTES|52=20201007-06:50:01.658|56=200171|55=AUDUSD.spa|262=200171_3|268=2|269=0|270=0.71329|269=1|270=0.71335|10=132|8=FIX.4.4|9=125|35=W|34=94|49=QUOTES|52=20201007-06:50:01.813|56=200171|55=EURCAD.spa|262=200171_5|268=2|269=0|270=1.56368|269=1|270=1.56389|10=130|8=FIX.4.4|9=124|35=W|34=95|49=QUOTES|52=20201007-06:50:01.814|56=200171|55=EURUSD.spa|262=200171_9|268=2|269=0|270=1.17514|269=1|270=1.1752|10=097|8=FIX.4.4|9=124|35=W|34=96|49=QUOTES|52=20201007-06:50:01.816|56=200171|55=EURAUD.spa|262=200171_4|268=2|269=0|270=1.64738|269=1|270=1.6475|10=094|8=FIX.4.4|9=126|35=W|34=97|49=QUOTES|52=20201007-06:50:01.870|56=200171|55=GBPJPY.spa|262=200171_11|268=2|269=0|270=136.409|269=1|270=136.426|10=190|8=FIX.4.4|9=125|35=W|34=98|49=QUOTES|52=20201007-06:50:01.903|56=200171|55=AUDUSD.spa|262=200171_3|268=2|269=0|270=0.71329|269=1|270=0.71336|10=131|';
  final msgRegex = RegExp(r'[^0-9a-zA-Z:\s]*8=FIX(.*?)[^0-9]10=\d\d\d.?');
  var matches = msgRegex.allMatches(messages);
  if (matches != null) {
    for (var match in matches) {
      print(match.group(0));
    }
  }
}


lib/src/fix_dart_base.dart
import 'dart:convert';
import 'dart:io';

class FixClient {
  final _tag = 'FixClient';
  final _protocol = 'FIX.4.4';
  Socket _socket;
  int _reqId = 1;

  String host, senderCompID, targetCompID, username, password;
  int port;

  FixClient(
      {this.host,
      this.port,
      this.senderCompID,
      this.targetCompID,
      this.username,
      this.password});

  bool get isConnected => _socket != null;

  ///Subscribe market data quote
  void subscribeQuote(String symbol) {
    if (isConnected) {
      var message = [
        '8=$_protocol',
        '35=V', //Message Type
        '262=$_reqId', //MDReqID
        '263=1', //SubscriptionRequestType
        '264=0', //MarketDepth
        '265=0', //MDUpdateType
        '267=2', //NoMDEntryTypes
        '146=1', //NoRelatedSym
        '55=$symbol', //Symbol
      ];
      _socket.add(utf8.encode(message.join(' ')));
      _reqId++;
    } else {
      print(
          '${DateTime.now()} $_tag can not subscribe quote, socket is not connected');
    }
  }

  ///Connect to FIX api
  void connect({
    bool autoReconnect = false,
    Function onConnected,
  }) {
    Socket.connect(host, port).then((socket) {
      print('${DateTime.now()} $_tag is connected to $host');
      _socket = socket;
      void dataHandler(data) {
        print(data);
        try {
          var parsed = {};
          for (var kv in '$data'.split(' ')) {
            var split = kv.split('=');
            parsed[split[0]] = parsed[1];
          }
          print(parsed);
          if (parsed['35'] == 'W') {
            // var symbol = parsed['55'];
            // var priceType = parsed['269'];
            // var price = parsed['270'];
            // var volume = parsed['271'];
            // var transactTime = parsed['60'];
            // var sendingTime = parsed['52'];
          } else if (parsed['35'] == 'Y') {
            print(
                '${DateTime.now()} $_tag market data rejected ${parsed['58']}');
          } else if (parsed['35'] == 'j') {
            print(
                '${DateTime.now()} $_tag business message rejected ${parsed['58']}');
          }
        } catch (e) {
          print('${DateTime.now()} $_tag failed to parse message $data');
        }
      }

      void errorHandler(e, trace) {
        print('${DateTime.now()} $_tag connection error $e');
      }

      void doneHandler() {
        print('${DateTime.now()} $_tag socket connection done');
        socket.destroy();
        _socket = null;
        _reqId = 1;
        if (autoReconnect) connect(autoReconnect: autoReconnect);
      }

      var logonMessage = [
        '8=$_protocol',
        '35=A', //Message Type
        '98=0', //Encryption method
        '108=30', //Heart Beat Interval
        '141=Y', //Reset Sequence Number Flag
        '553=$username', //Username
        '554=$password', //Password
      ];
      socket.add(utf8.encode(logonMessage.join(' ')));

      if (onConnected != null) onConnected();

      socket.listen(
        dataHandler,
        onError: errorHandler,
        onDone: doneHandler,
        cancelOnError: false,
      );
    }).catchError((e) {
      _socket = null;
      _reqId = 1;
      print('${DateTime.now()} $_tag can not connect to host $e');
      if (autoReconnect) connect(autoReconnect: autoReconnect);
    });
  }
}


lib/src/fix_client.dart
import 'dart:async';
import 'dart:convert';
import 'dart:io';

import 'package:fix_dart/src/fix_quote.dart';

import 'fix_44.dart';

typedef OnQuoteReceived = Function(Quote quote);

class FixDart {
  final _soh = String.fromCharCode(1);
  final _separator = '|';
  final _tag = 'FixClient';
  final _version = 'FIX.4.4';
  final msgRegex = RegExp(r'[^0-9a-zA-Z:\s]*8=FIX(.*?)[^0-9]10=\d\d\d.?');
  Socket _socket;
  int _reqId = 1;
  var isConnecting = true;
  Timer _hearbeat;
  bool autoReconnect = true;
  Function(String log) onWriteLog;

  String host, senderCompID, targetCompID, username, password;
  int port;

  bool debug = false;

  FixDart(
      {this.host,
      this.port,
      this.senderCompID,
      this.targetCompID,
      this.username,
      this.password});

  bool get isConnected => _socket != null;
  Socket get connection => _socket;

  ///Subscribe market data
  void subscribeMarketData(String symbol, [int depth = 0]) {
    if (isConnected) {
      var msg = serialize(
        MsgTypes.MARKETDATAREQUEST,
        [
          MDReqID(_reqId),
          SubscriptionRequestType(1),
          MarketDepth(depth),
          MDUpdateType(0),
          NoMDEntryTypes(2),
          MDEntryType(0),
          MDEntryType(1),
          NoRelatedSym(1),
          Symbol(symbol)
        ],
      );
      _socket.add(utf8.encode(msg));
      if (debug) printLog('⤴️ ${msg.replaceAll(_soh, _separator)}');
      _reqId++;
    } else {
      printLog('🔕 can not subscribe market data, socket is not connected');
    }
  }

//Serialize fix message
  String serialize(MsgTypes msgType, List<FixTag> body) {
    var header = [
      MsgType(msgType),
      SendingTime(DateTime.now()
          .toUtc()
          .toIso8601String()
          .replaceAll('-', '')
          .replaceAll('T', '-')
          .replaceAll('Z', '')
          .substring(0, 21)),
      SenderCompID(senderCompID),
      TargetCompID(targetCompID),
      MsgSeqNum(_reqId),
    ];
    var headerMessage = header.map((e) => e.toKV()).join(_soh);
    var bodyMessage = body.map((e) => e.toKV()).join(_soh);
    var sepCount = bodyMessage.isNotEmpty ? 2 : 1;
    var out = [
      BeginString(_version).toKV(),
      BodyLength(headerMessage.length + bodyMessage.length + sepCount).toKV(),
      headerMessage,
      if (bodyMessage.isNotEmpty) bodyMessage,
    ];
    var output = '${out.join(_soh)}$_soh';
    var sum = utf8.encode(output).reduce((a, b) => a + b);
    var checksum = (sum % 256).toString().padLeft(3, '0');
    output = '$output${CheckSum(checksum).toKV()}$_soh';
    return output;
  }

  ///Connect to FIX api
  void connect({
    bool autoReconnect = false,
    int reconnectDelay = 0,
    Function onConnected,
    OnQuoteReceived onQuoteReceived,
  }) {
    this.autoReconnect = autoReconnect;
    isConnecting = true;
    Socket.connect(host, port).then((socket) async {
      isConnecting = false;
      _socket = socket;
      _socket.setOption(SocketOption.tcpNoDelay, true);
      printLog('🔀 is connected to $host:$port');
      _hearbeat = Timer.periodic(Duration(seconds: 30), (t) {
        //Send heartbeat message
        var msg = serialize(
          MsgTypes.HEARTBEAT,
          [
            TestReqID(_reqId),
          ],
        );
        _socket.add(utf8.encode(msg));
        if (debug) printLog('⤴️ ${msg.replaceAll(_soh, _separator)}');
        _reqId++;
      });
      void dataHandler(event) {
        try {
          var raw = utf8.decode(event).replaceAll(_soh, _separator);
          if (debug) printLog('⤵️ $raw');

          var matches = msgRegex.allMatches(raw);
          if (matches != null) {
            for (var match in matches) {
              var data = match.group(0);
              var parsed = {};
              for (var kv in '$data'.split(_separator)) {
                var split = kv.split('=');
                if (split.length > 1) {
                  var k = split[0];
                  var v = split[1];
                  if (parsed[k] != null) {
                    parsed['$k$k'] = v;
                  } else {
                    parsed[k] = v;
                  }
                }
              }
              // print(parsed);
              var serverMessage = parsed[Text().tag];
              var messageType = parsed[MsgType().tag];
              if (serverMessage != null) {
                printLog('💬 $serverMessage');
              }
              //HEARTBEAT
              if (messageType == '0') {
              }
              //MARKETDATASNAPSHOTFULLREFRESH
              else if (messageType == 'W') {
                var symbol = parsed[Symbol().tag];
                var sendingTime = DateTime.tryParse(
                    '${parsed[SendingTime().tag]}Z'.replaceFirst('-', ' '));
                var BID = MDEntryPx().tag;
                var OFFER = '$BID$BID';
                var ask = double.tryParse('${parsed[OFFER]}');
                var bid = double.tryParse('${parsed[BID]}');
                var quote = Quote(symbol, ask, bid, sendingTime);
                if (onQuoteReceived != null) onQuoteReceived(quote);
              }
            }
          }
        } catch (e, t) {
          print('${DateTime.now()} $_tag failed to parse message $e\n$t');
        }
      }

      void errorHandler(e, trace) {
        _reqId = 1;
        printLog('⛔️ connection error $e');
      }

      void doneHandler() async {
        printLog(
            '🚫 socket connection done, will reconnect? ${this.autoReconnect}');
        _socket.destroy();
        _socket = null;
        _reqId = 1;
        try {
          _hearbeat?.cancel();
        } catch (e) {
          //
        }
        if (this.autoReconnect) {
          await Future.delayed(Duration(seconds: reconnectDelay), () {
            if (!isConnecting && !isConnected) {
              connect(
                autoReconnect: this.autoReconnect,
                reconnectDelay: reconnectDelay,
                onConnected: onConnected,
                onQuoteReceived: onQuoteReceived,
              );
            }
          });
        }
      }

      _socket.listen(
        dataHandler,
        onError: errorHandler,
        onDone: doneHandler,
        cancelOnError: true,
      );

      //Send logon message
      var msg = serialize(
        MsgTypes.LOGON,
        [
          EncryptMethod(0),
          HeartBtInt(30),
          ResetSeqNumFlag('Y'),
          Username(username),
          Password(password)
        ],
      );
      _socket.add(utf8.encode(msg));
      _reqId++;
      if (debug) printLog('⤴️ ${msg.replaceAll(_soh, _separator)}');
      if (onConnected != null) onConnected();
    }).catchError((e) {
      _socket = null;
      _reqId = 1;
      print('${DateTime.now()} $_tag can not connect to host $e');
      if (this.autoReconnect) {
        connect(autoReconnect: this.autoReconnect, onConnected: onConnected);
      }
    });
  }

  void printLog(String message) {
    print('${DateTime.now()} $_tag $message');
    if (onWriteLog != null) onWriteLog(message);
  }
}
